pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+	if len(address.Hostname) > 0 {
+		return address.Hostname, true
+	}
+	if hostRecord, exists := podHostnames[address.IP]; exists && validation.IsDNS1123Label(hostRecord.HostName) {
+		return hostRecord.HostName, true
+	}
+	return "", false
+}
+
+func getPodHostnamesFromAnnotation(annotations map[string]string) (map[string]endpoints.HostRecord, error) {
+	hostnames := map[string]endpoints.HostRecord{}
+
+	if annotations != nil {
+		if serializedHostnames, exists := annotations[endpoints.PodHostnamesAnnotation]; exists && len(serializedHostnames) > 0 {
+			err := json.Unmarshal([]byte(serializedHostnames), &hostnames)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return hostnames, nil
+}
+
+func (kd *KubeDNS) generateSRVRecordValue(svc *kapi.Service, portNumber int, cNameLabels ...string) *skymsg.Service {
+	cName := strings.Join([]string{svc.Name, svc.Namespace, serviceSubdomain, kd.domain}, ".")
+	for _, cNameLabel := range cNameLabels {
+		cName = cNameLabel + "." + cName
+	}
+	recordValue, _ := getSkyMsg(cName, portNumber)
+	return recordValue
+}
+
+// Generates skydns records for a headless service.
+func (kd *KubeDNS) newHeadlessService(service *kapi.Service) error {
+	// Create an A record for every pod in the service.
+	// This record must be periodically updated.
+	// Format is as follows:
+	// For a service x, with pods a and b create DNS records,
+	// a.x.ns.domain. and, b.x.ns.domain.
+	key, err := kcache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		return err
+	}
+	e, exists, err := kd.endpointsStore.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to get endpoints object from endpoints store - %v", err)
+	}
+	if !exists {
+		glog.V(1).Infof("Could not find endpoints for service %q in namespace %q. DNS records will be created once endpoints show up.", service.Name, service.Namespace)
+		return nil
+	}
+	if e, ok := e.(*kapi.Endpoints); ok {
+		return kd.generateRecordsForHeadlessService(e, service)
+	}
+	return nil
+}
+
+func (kd *KubeDNS) Records(name string, exact bool) ([]skymsg.Service, error) {
+	glog.Infof("Received DNS Request:%s, exact:%v", name, exact)
+	trimmed := strings.TrimRight(name, ".")
+	segments := strings.Split(trimmed, ".")
+	path := reverseArray(segments)
+	if kd.isPodRecord(path) {
+		response, err := kd.getPodRecord(path)
+		if err == nil {
+			return []skymsg.Service{*response}, nil
+		}
+		return nil, err
+	}
+
+	if exact {
+		key := path[len(path)-1]
+		if key == "" {
+			return []skymsg.Service{}, nil
+		}
+		if record, ok := kd.cache.GetEntry(key, path[:len(path)-1]...); ok {
+			return []skymsg.Service{*(record.(*skymsg.Service))}, nil
+		}
+		return nil, etcd.Error{Code: etcd.ErrorCodeKeyNotFound}
+	}
+
+	// tmp, _ := kd.cache.Serialize("")
+	// glog.Infof("Searching path:%q, %v", path, tmp)
+	records := kd.cache.GetValuesForPathWithRegex(path...)
+	retval := []skymsg.Service{}
+	for _, val := range records {
+		retval = append(retval, *(val.(*skymsg.Service)))
```

Comment:
I think you should validate the type assertion here. Something like this:

```
rec, ok := val.(*skymsg.Service)
if !ok {
   return fmt.Errorf("invalid record type received: %T", val)
}
retval = append(retval, rec)
```

If something goes wrong and you get an object of unexpected type, Go panics.